File: src/bin/0973-k-closest-points-to-origin.py
# @leetup=info id=973 lang=python3 slug=k-closest-points-to-origin



# @leetup=code

import heapq


class Solution:
    def kClosest(self, points: list[list[int]], k: int) -> list[list[int]]:
        q = []
        for p in points:
            # heapq.heappush(q, (-(p[0] * p[0] + p[1] * p[1]), p))
            # if len(q) > k:
            #     heapq.heappop(q)

            # Use heappushpop
            if len(q) == k:
                heapq.heappushpop(q, (-(p[0] * p[0] + p[1] * p[1]), p))
            else:
                heapq.heappush(q, (-(p[0] * p[0] + p[1] * p[1]), p))

        return [p for (_, p) in q]


# @leetup=code

File: src/bin/0973-k-closest-points-to-origin.rs
// @leetup=info id=973 lang=rust slug=k-closest-points-to-origin

// @leetup=code
use std::collections::*;

impl Solution {
    pub fn k_closest(points: Vec<Vec<i32>>, k: i32) -> Vec<Vec<i32>> {
        let mut q = BinaryHeap::new();
        for p in points {
            q.push((p[0] * p[0] + p[1] * p[1], p));
            if q.len() > k as usize {
                q.pop();
            }
        }
        q.into_iter().map(|(_, p)| p).collect()
    }
}
// @leetup=code

struct Solution;

fn main() {
    let _ = Solution::k_closest;
}
